packages/next/src/server/base-server.ts

```diff
@@ -3345,6 +3351,18 @@ export default abstract class Server<
         delete headers[NEXT_CACHE_TAGS_HEADER]
       }
 
+      // If cache control is already set on the response we don't
+      // override it to allow users to customize it via next.config
+      if (
+        typeof cacheEntry.revalidate !== 'undefined' &&
+        !res.getHeader('Cache-Control')
+      ) {
+        headers['Cache-Control'] = formatRevalidate({
+          revalidate: cacheEntry.revalidate,
+          expireTime: this.nextConfig.expireTime,
+        })
+      }
```

Comment:
This is the major fix, as revalidate is only set when the route is being statically generated.